source/common/filter/tcp_proxy.cc

```diff
@@ -162,8 +162,14 @@ void TcpProxy::finalizeUpstreamConnectionStats() {
 
 void TcpProxy::closeUpstreamConnection() {
   finalizeUpstreamConnectionStats();
-  upstream_connection_->close(Network::ConnectionCloseType::NoFlush);
-  read_callbacks_->connection().dispatcher().deferredDelete(std::move(upstream_connection_));
+
+  // Move the upstream_connection_ into a temporary so that when
+  // onUpstreamEvent() is called with LocalClose, we know that it is
+  // already closed, and don't go into an infinite event loop.
+  Network::ClientConnectionPtr conn(std::move(upstream_connection_));
```

Comment:
Yikes, infinite loops!  I checked with Greg to understand this better and I find it really strange that the connection class can raise multiple Close calls.  @mattklein123 is it worth filing an open issue that we should address this?    I think this is a reasonable short term workaround for what would be a much higher risk change but maybe there's a good reason for multiple Close calls that I'm missing :-)

Reply:
It shouldn't raise multiple closes AFAIK. @ggreenway how is that happening?

Reply:
hahaha, I've been making my own life hard. MockConnection will raise multiple types of close (if you use some of it's default behaviors). ConnectionImpl is doing the right thing.

I'll take another pass on this and make MockConnection do the right thing for these tests, and then this line should go away.

Reply:
Nope, I was way off.

It turns out the complication was because there was a function closeUpstreamConnection(), which would call upstream_connection_->close() (among other things), and this function was called from onUpstreamEvent() when the connection is already being closed. That cycle in the callgraph was making everything more difficult, so I reworked it all to remove that. PTAL at the latest.